### PR TITLE
feat(SCIM): optimize performance

### PR DESCRIPTION
**Pull request by danflomin**
In this PR the main changes relate to using async overloads of DB queries.
Using async overloads is considered as the good practice in asp.net, and after a benchmark I conducted, these changes improved the performance by several magnitudes.

Also, the usage of `yield` in DB queries is considered to be less efficient with respect to network.

Another performance improvements introduced in here: 
* The `ResolveChildren` and `ResolveParents` functions, queried the same representations multiple times, so I added a check that makes sure that each representation is queried only once.
* The implementation of GET methods can use the `AsNoTracking` behavior which improves performance.
* I used the paginated variation in a few places that did not use it.

### CODE REVIEW

File: src/Scim/SimpleIdServer.Scim.Persistence.EF/Extensions/EFSCIMExpressionLinqExtensions.cs [hunk 1]
```diff
@@ -43,12 +42,10 @@ public static class EFSCIMExpressionLinqExtensions
             {
                 filteredAttrs = filteredAttrs.Where(a => a.RepresentationId == id);
                 var result = await representations.FirstOrDefaultAsync(r => r.Id == id && r.ResourceType == resourceType);
-                var includedFullPathLst = (includedAttributes != null && includedAttributes.Any()) ? includedAttributes.Where(i => i is SCIMComplexAttributeExpression).Select(i => i.GetFullPath()) : new List<string>();
                 result.FlatAttributes = filteredAttrs.ToList();
                 return result;
             }
-
-            representations = representations.Include(r => r.FlatAttributes).ThenInclude(s => s.SchemaAttribute);
```
Comment (danflomin) [inline src/Scim/SimpleIdServer.Scim.Persistence.EF/Extensions/EFSCIMExpressionLinqExtensions.cs:51]: This was moved to `EFSCIMRepresentationQueryRepository`

File: src/Scim/SimpleIdServer.Scim/Helpers/RepresentationReferenceSync.cs [hunk 2]
```diff
@@ -20,32 +17,32 @@ public class RepresentationReferenceSync : IRepresentationReferenceSync
...
+			ISCIMSchemaCommandRepository scimSchemaCommandRepository)
         {
 			_scimAttributeMappingQueryRepository = scimAttributeMappingQueryRepository;
 			_scimRepresentationCommandRepository = scimRepresentationCommandRepository;
 			_scimSchemaCommandRepository = scimSchemaCommandRepository;
-			_resourceTypeResolver = resourceTypeResolver;
 		}
 
-		public virtual IEnumerable<RepresentationSyncResult> Sync(string resourceType, SCIMRepresentation oldScimRepresentation, SCIMRepresentation newSourceScimRepresentation, string location, SCIMSchema schema, bool updateAllReferences = false, bool isScimRepresentationRemoved = false)
+        public virtual async Task<List<RepresentationSyncResult>> Sync(string resourceType, SCIMRepresentation oldScimRepresentation, SCIMRepresentation newSourceScimRepresentation, string location, SCIMSchema schema, bool updateAllReferences = true, bool isScimRepresentationRemoved = false)
```
Comment (danflomin) [inline src/Scim/SimpleIdServer.Scim/Helpers/RepresentationReferenceSync.cs:31]: I rebased my code which originated from 4.0.2 using `master`, and I see that you made some changes like `false` -> `'true`.
@thabart should I change this?

File: src/Scim/SimpleIdServer.Scim/Helpers/RepresentationReferenceSync.cs [hunk 3]
```diff
@@ -54,67 +51,64 @@ public IEnumerable<RepresentationSyncResult> Sync(IEnumerable<SCIMAttributeMappi
...
                 }
 
                 foreach (var attributeMapping in kvp.Where(r => !r.IsSelf))
                 {
                     result = new RepresentationSyncResult();
-                    var targetSchema = mappedSchemas.First(s => s.ResourceType == attributeMapping.TargetResourceType && s.Attributes.Any(a => a.Id == attributeMapping.TargetAttributeId));
+                    var targetSchema = targetSchemas.First(s => s.ResourceType == attributeMapping.TargetResourceType);
                     var targetAttribute = targetSchema.GetAttributeById(attributeMapping.TargetAttributeId);
-                    var sourceAttribute = sourceSchema.GetAttributeById(attributeMapping.SourceAttributeId);
+                    var sourceAttribute = schema.GetAttributeById(attributeMapping.SourceAttributeId);
```
Comment (danflomin) [inline src/Scim/SimpleIdServer.Scim/Helpers/RepresentationReferenceSync.cs:70]: same

**Comment by simpleidserver** [comment]
Hello,

Firstly, thank you for your pull request.

Are you sure that the yield clause is less efficient in DB queries? I checked with and without the yield keyword, and the number of RPC requests made against the SQLSERVER is the same. It should also be the same with a POSTGRESQL database. In the pull request, all the data is stored in-memory, which can be problematic if there are many references to update.

I'm okay with using .NET version 7, but please remember to upgrade the other projects that reference this project.

Unfortunately, the solution is not compiling. Could you please fix the issues?

Thank you.

File: src/Scim/SimpleIdServer.Scim/Helpers/RepresentationReferenceSync.cs [hunk 4]
```diff
@@ -54,67 +51,64 @@ public IEnumerable<RepresentationSyncResult> Sync(IEnumerable<SCIMAttributeMappi
...
                     if (duplicateIds.Any()) throw new SCIMUniquenessAttributeException(string.Format(Global.DuplicateReference, string.Join(",", duplicateIds.Select(_ => _.Key).Distinct())));
                     newIds = newIds.Except(oldIds).ToList();
-                    existingIds = oldIds.Except(idsToBeRemoved);
                 }
 
                 foreach (var attributeMapping in kvp.Where(r => !r.IsSelf))
                 {
                     result = new RepresentationSyncResult();
-                    var targetSchema = mappedSchemas.First(s => s.ResourceType == attributeMapping.TargetResourceType && s.Attributes.Any(a => a.Id == attributeMapping.TargetAttributeId));
+                    var targetSchema = targetSchemas.First(s => s.ResourceType == attributeMapping.TargetResourceType);
```
[the remark this answers is not in the document]
Comment (simpleidserver) [reply src/Scim/SimpleIdServer.Scim/Helpers/RepresentationReferenceSync.cs:68]: "This portion of code has been updated by pull request #557.
There was an issue during the resolution of the attribute present in a custom schema.
You are welcome to retain this change :)

File: src/Scim/SimpleIdServer.Scim/Helpers/RepresentationReferenceSync.cs (hunk 3 above)
Comment (simpleidserver) [reply src/Scim/SimpleIdServer.Scim/Helpers/RepresentationReferenceSync.cs:70]: Same remark

File: src/Scim/SimpleIdServer.Scim/Helpers/RepresentationReferenceSync.cs (hunk 2 above)
Comment (simpleidserver) [reply src/Scim/SimpleIdServer.Scim/Helpers/RepresentationReferenceSync.cs:31]: I checked the `master` branch, by default, the field `updateAllReferences` is equals to `false` and not to `true` : 

https://github.com/simpleidserver/SimpleIdServer/blob/ed077178cd645a92c4aab502b4266b3f923313fb/src/Scim/SimpleIdServer.Scim/Helpers/RepresentationReferenceSync.cs#L18

The field must be `false`

**Comment by danflomin** [comment]
Hello @simpleidserver 

1. Yes - I'm sure it is less efficient.
2. Eventually we will have all the references in the handlers anyway, right?
3. The change I made in the `ResolveChildren` and `ResolveParents` actually solved an OOM exception I previously got.
4. I updated the PR and the build now works.

Kind regards
Dan

**Comment by simpleidserver** [comment]
Unfortunately, the unit tests are not functioning properly :(
Could you please revert your changes to the TargetFramework or modify the unit test projects accordingly?

Additionally, for my curiousity :), do-you have a link to documentation that explains why the `yield` keyword should not be used during database querying. As far as my understanding goes, when the `yield` statement is employed, the compiler transforms it into a state machine that implements `IEnumerable`. Subsequently, the foreach loop invokes the `MoveNext` method to retrieve results.

When `MoveNext` is invoked, it is essential to subsequently call `ToList`; otherwise, the SQL query will be executed by the calling application within the foreach loop.

For example, consider the previous code snippet, `FindPaginatedGraphAttributes` :


```
public IEnumerable<IEnumerable<SCIMRepresentationAttribute>> FindPaginatedGraphAttributes(string valueStr, string schemaAttributeId, int nbRecords = 50, string sourceRepresentationId = null)
{
    var query = _scimDbContext.SCIMRepresentationAttributeLst.AsNoTracking()
        .Where(a => a.SchemaAttributeId == schemaAttributeId && a.ValueString == valueStr || (sourceRepresentationId != null && a.ValueString == sourceRepresentationId))
        .OrderBy(r => r.ParentAttributeId)
        .Select(r => r.ParentAttributeId)
        .AsNoTracking();
    var nb = query.Count();
    var nbPages = Math.Ceiling((decimal)(nb / nbRecords));
    for (var i = 0; i <= nbPages; i++)
    {
        var parentIds = query.Skip(i * nbRecords).Take(nbRecords);
        var result = _scimDbContext.SCIMRepresentationAttributeLst.AsNoTracking()
            .Where(a => parentIds.Contains(a.Id) || parentIds.Contains(a.ParentAttributeId));
        yield return result;
    }
}
```

It should be updated to something like this : 

``` 
public IEnumerable<IEnumerable<SCIMRepresentationAttribute>> FindPaginatedGraphAttributes(string valueStr, string schemaAttributeId, int nbRecords = 50, string sourceRepresentationId = null)
{
    var query = _scimDbContext.SCIMRepresentationAttributeLst.AsNoTracking()
        .Where(a => a.SchemaAttributeId == schemaAttributeId && a.ValueString == valueStr || (sourceRepresentationId != null && a.ValueString == sourceRepresentationId))
        .OrderBy(r => r.ParentAttributeId)
        .Select(r => r.ParentAttributeId)
        .AsNoTracking();
    var nb = query.Count();
    var nbPages = Math.Ceiling((decimal)(nb / nbRecords));
    for (var i = 0; i <= nbPages; i++)
    {
        var parentIds = query.Skip(i * nbRecords).Take(nbRecords);
        var result = _scimDbContext.SCIMRepresentationAttributeLst.AsNoTracking()
            .Where(a => parentIds.Contains(a.Id) || parentIds.Contains(a.ParentAttributeId));
        yield return result.ToList();
    }
}
```

**Comment by simpleidserver** [comment]
Hello @danflomin ,

I fixed the UTs and also issues in the `RepresentationReferenceSync` class.
If you are agree with the changes, i'll merge them into the branch.

KR,

SID

File: src/Scim/SimpleIdServer.Scim/SimpleIdServer.Scim.csproj [hunk 5]
```diff
@@ -2,11 +2,13 @@
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
     <Description>SCIM2.0 implementation.</Description>
+	<LangVersion>latest</LangVersion>
```
Comment (danflomin) [inline src/Scim/SimpleIdServer.Scim/SimpleIdServer.Scim.csproj:5]: why is this needed?

**Comment by danflomin** [comment]
Hello 

I will try to find a good explanation for the issue with the yield.

Thank you for your fixes, I went over most of it and it looks fine.
I still need to understand the changes in `RepresentationReferenceSync` and in `RepresentationSyncResult`.

I will let you know once done.

Kind regards 
Dan

File: src/Scim/SimpleIdServer.Scim/SimpleIdServer.Scim.csproj (hunk 5 above)
Comment (simpleidserver) [reply src/Scim/SimpleIdServer.Scim/SimpleIdServer.Scim.csproj:5]: Because some changes in the pull request are not supported by previous c# version for example `??=`

**Comment by gabrielemilan** [comment]
Hello, do you have ETA of this PR?

Kind regards.
Gabriele
